### ISSUE TITLE
Hyperband stopping criterion too permissive with noisy metrics

### ISSUE THREAD

**Issue by leopd** [wandb/client#1234]
`wandb --version && python --version && uname`

* Weights and Biases version: 0.9.7
* Python version: 3.7.9
* Operating System: Ubuntu 18.04LTS

### Description

Running a sweep using Hyperband (HB) early stopping criterion.  The metric I use to evaluate model quality is somewhat noisy.  The current hyperband stopping criterion considers the run's minimum (best) [(code)](https://github.com/wandb/client/blob/63f3d7806669d2dd8d459134c2f88d6e341a797e/wandb/sweeps/hyperband_stopping.py#L135) metric value vs a single-sample snapshot of the other runs in that band. [(code)](https://github.com/wandb/client/blob/63f3d7806669d2dd8d459134c2f88d6e341a797e/wandb/sweeps/hyperband_stopping.py#L102).  For noisy metrics, the result is that runs are judged very favorably, and runs that a human would obviously judge as being worse than average are allowed to continue.  This problem becomes especially pronounced in later bands where the `min()` is taken over a larger number of samples.

### What I Did

Consider these charts of the objective metric vs time for a HB run.   

![smoothed objective metrics](https://user-images.githubusercontent.com/193183/92950266-3d66f380-f411-11ea-96cf-9105d884822b.png)

The run I'm concerned about is in hot pink, with the badly drawn label "This V".  It passed a band marker at about 3.7 hours where the two brownish runs at the bottom finished.  This smoothed plot makes it clear that the hot pink run is not in the better half of all runs to reach that point, and thus should be stopped.  But because it got one lucky sample below the median, it is allowed to continue, as shown in this unsmoothed plot:

![unsmoothed objective metrics](https://user-images.githubusercontent.com/193183/92949691-5ae78d80-f410-11ea-8c82-dce216f4191c.png)

### Discussion

It's not totally clear what to do about this.  But the current criterion is clearly biased in a way that leads to wasted compute because it stops a lot fewer runs than it should when the metrics are noisy.  Using single sample snapshots for both the current run and the thresholds I think would yield an unbiased criterion with high variance.  Alternately using the same aggregation for both could also eliminate the bias.  I'm not sure which would do better in general.

**Comment by cvphelps** [comment]
Hi there, thanks for the feedback on the hyperband implementation! We based it on the original hyperband paper — if you have any ideas on what we could improve here to make our algorithm more robust to noisy metrics, definitely interested in making this better for you :)

**Comment by leopd** [comment]
That [original paper](https://arxiv.org/abs/1603.06560) is a monster, weighing in at 33 pages before appendices and references.  It's long on proofs and short on practicalities.  So I certainly sympathize with anybody trying to build software based on it -- it is far from a model for clarity, even though the key ideas they propose are brilliant in their simplicity.  Given that wandb exposes the idyllic combination of BayesOpt  + Hyperband, I recommend [the BOHB paper by Falkner, Klein, and Hutter](https://arxiv.org/pdf/1807.01774.pdf) which dives into some subtleties on getting them to work together as that original paper leaves out a lot of details, forcing us as builders to make our best educated guesses about how to achieve what they describe.

Quoting the original paper, which is surprisingly clear in section 3.1 on the key idea of successive halving: "uniformly allocate a budget to a set of hyperparameter configurations, evaluate the performance of all configurations, **throw out the worst half**, and repeat until one configuration remains." (Emphasis mine.)  The bug I'm raising here is that the current code is often throwing out less than half of the configurations that pass each band, because it's biased by the different ways of sampling / aggregating the runs' metric histories.  This can result in a significant decrease in HPO performance (compute time to achieve a desired quality).

**Comment by leopd** [comment]
Sorry - I realize a mistake I made in the original bug report.  I shouldn't have referred to human judgement, because that implies that the algorithm is doing something silly that an expert wouldn't - this isn't the problem.  The algorithm as implemented is objectively incorrect.

At its core, the problem is that it's making an unfair comparison - a single metric vs the min of a series of metrics. The bias means it fails to stop half (or 1/eta) the jobs like it should.  The question is how to fix it.  Compare single value against single value at the instant it reaches the new band?  That would be fair, and simple, but also very noisy.  It achieves the key goal of stopping half the runs, and so all the very long proofs in the paper would apply, but the variance would be very high, meaning that the convergence to a good configuration would be slow.  A lower variance alternative would be to aggregate the metrics within the band in the same way for every run.  Min vs min could work, but that will also have high variance (higher?), as it emphasizes outliers.

I think mean is probably better, since it will lead to the lowest variance assuming normal noise.  Of course these samples are expected to have some systematic progression within the band towards better metric values.  Still, the mean will reduce variance and will be as fair a comparison between runs as anything.  When I say the metrics within the band, I mean just the new metrics since the end of the last band.  So if you band boundaries are {2,4,8}, then when aggregating metrics for band=8, you'd use the metrics at times 5 through 8.  No reason to include the earlier history, since the runs already passed that criterion.

Here's another example where the current implementation will fail badly, and applies even if the objective metric has zero noise: if the training processes start to overfit, such that the metrics of all runs bottom out and then start to go back up (get worse).  In this situation, the current algorithm would never stop a large class of runs which have started to overfit.  Because it would always consider the best metric the run had ever achieved, and compare that against the larger overfitting metric of all the other runs (including itself) which are going back up in the later bands.  The minimum metric value is going to be better than what all the runs are doing later, and so the current implementation will judge every run favorably and never stop them.

**Comment by SamPruden** [comment]
@leopd's writeup of this issue is excellent. Is there any progress on this @cvphelps? I'm trying to work out if I should build around the WandB sweeping or the beta Ray Tune sweeping, which I'm hoping won't have this problem.

@leopd, just to add to your suggestions about using the mean instead, would a smoothed/lowpassed comparison at the comparison point maybe be a good option? Equivalently, a weighted average with a recency bias.

**Comment by lukas** [comment]
Sorry - I totally missed this and it really should compare the snapshot - let me take a look.

The mean or a smoothed option seems like a good idea, although maybe simpler and more flexible is to let the user do the aggregation. Thanks @leopd.

**Comment by SamPruden** [comment]
> Sorry - I totally missed this and it really should compare the snapshot - let me take a look.
> 
> The mean or a smoothed option seems like a good idea, although maybe simpler and more flexible is to let the user do the aggregation. Thanks @leopd.

I'm glad this is being looked at! Giving the user control of the aggregation sounds great if there's a sensible default too. Implementing and tuning a custom "minimum of smoothed" might be slightly fiddly as a user, so it would be nice if there were a utility provided to make that simple. Perhaps if the smoothing amount were parameterised the same way as the smoothing in the UI, it would be easy to fiddle with a slider to pick a smoothing value that looks right.

**Comment by leopd** [comment]
@SamPruden I think some kind of smoothing criterion makes sense - without a rigorous analysis, intuitively it should reduce variance in the quality of tuning.  It's essentially impossible to know what distribution the noise looks like, so we should just assume normal, in which case using the mean is probably the best smoother.  Over what range?  I'd say only the most recent band.  If you take the mean over the entire run so far, then you'll frequently pull in extremely outliers from the very beginning of the run (where the quality is horrible because the net was just randomly init'd).  Also, if your training process produces learning curves that are non-monotonic (pretty common) then an aggregator like min() will have some pathological behavior.

Re: exposing aggregation options to the user - I'd advise against it.  Of course there's no best "right" answer here (barring a full research paper's worth of effort to test), so leaving it as a hyper-hyper-parameter seems reasonable.  But the key reason for tuning is to automate finding good hyper-parameters, and every time you expose hyper-hyper-parameters, you run the risk that somebody thinks they need to do multiple sweeps varying that hyper-hyper-parameter and then they're totally wasting their time and money.  Those extra lines in the documentation to describe the option will make a few people slightly happier (although maybe they shouldn't be), and everybody else slightly more confused.

**Comment by SamPruden** [comment]
My thought on exposing aggregation options was that it would be nice to have that power **if there are some sensible defaults**, but I completely agree that this shouldn't have to be a hyperparameter that most users need to tune, or spend time researching or thinking about.

My instinct is that "minimum of lowpass" would be a good option. It should deal with noise problems, but not be sensitive to the overall shape of the whole curve. Skipping early data could be done if necessary per @leopd's suggestion. @leopd I'm not sure what you mean by "mean" - if you mean moving average then I think that we're in agreement, but if you're talking about taking the mean over a large range then I worry that that would be too sensitive to the shape of the curve.

I haven't done more than skimmed the paper, but if it's true that it's not clear on what metrics they used to achieve their results, we could always just shoot the authors an email.

**Comment by SamPruden** [comment]
To motivate the idea that customisation may be good, different training scenarios may have different ideas of "best".

For example, if you take a fairly standard setup, then this may heavily depend on batch size. At one extreme, you may have many tiny batches and would want heavy smoothing. At the other extreme, you may have only a single large batch, and might want no smoothing at all - only a simple minimum.

These scenarios may be quite rare, but it would be a shame to make this feature so locked down that it's not usable in them. For that reason, I would advocate sensible defaults and the option to customise with a big "**WARNING: Only mess with this if you know what you're doing!**" in the docs.

Counterpoint: A user could always just log a custom metric that's intended to be used by hyperband, so actually we already have customisation power without exposing that option here. This might impose extra fiddly data tracking on the user, but would keep the API simple.
